Fix seek of PagedInputStream over uncompressed data

oerling opened the pull request:
PagedInputStream decompresses a whole compression entry and returns
windows into that. If the underlying data is not compressed, it
returns the windows it gets from its input. When seeking to a
position, if the target position falls in the current compression run,
we return a window into the uncompressed data starting at the desired
offset. This offset is anywhere between the start and end of the
current uncompressed buffer.

However, the compression run may be covered by multiple consecutive
windows into the input stream buffer.  If these windows of the are
directly returned to the caller, the seek to position will only work
for positions that fall in the current window, which may not include
the start of the compression run.

We add a check that we do not backup outside of the last window if
directly returning windows into the input's buffer. When seeking to
position, we distinguish the case of returning windows into the input
directly from the case of returning windows into the uncompressed
buffer.

mbasmanova commented on velox/dwio/dwrf/common/ByteRLE.cpp:
```diff
@@ -354,6 +354,8 @@ void ByteRleDecoder::seekToRowGroup(PositionProvider& positionProvider) {
   // force reading a new header
   remainingValues = 0;
+  static int counter = 0;
+  LOG(INFO) << "Suspect Seek #" << counter++;
```
Should this be removed?

mbasmanova commented on velox/dwio/dwrf/common/PagedInputStream.cpp:
```diff
@@ -188,6 +190,16 @@ void PagedInputStream::BackUp(int32_t count) {
       "Backup without previous Next in ",
       getName());
+  if (state_ == State::ORIGINAL) {
+    VELOX_CHECK(
```
nit: perhaps, split this into two checks so that we have a better idea which boundary is violated in case of failure

mbasmanova commented on velox/dwio/dwrf/common/PagedInputStream.h:
```diff
@@ -34,9 +34,11 @@ class PagedInputStream : public SeekableInputStream {
     DWIO_ENSURE(
         decompressor_ || decrypter_,
         "one of decompressor or decryptor is required");
+    LOG(INFO) << "Make " << counter++ << " = " << this;
```
Remove.

mbasmanova commented on velox/dwio/dwrf/common/PagedInputStream.h:
```diff
@@ -125,7 +127,12 @@ class PagedInputStream : public SeekableInputStream {
   size_t remainingLength_{0};
 
   // the last buffer returned from the input
```
Remove this comment?

mbasmanova commented on velox/dwio/dwrf/common/PagedInputStream.h:
```diff
@@ -125,7 +127,12 @@ class PagedInputStream : public SeekableInputStream {
   size_t remainingLength_{0};
 
   // the last buffer returned from the input
+  // The first byte. in the range from input_->Next()
```
typos: "byte. in the"

mbasmanova reviewed (approved):
@oerling Orri, thank you for fixing this issue. Some comments below.

oerling commented:
Thank you,  Masha. Fixed the comments. Will first try this and then land next week, fix the other items before land.

 

 

 

From: Maria Basmanova ***@***.***> 
Sent: Thursday, June 9, 2022 9:57 PM
To: facebookincubator/velox ***@***.***>
Cc: oerling ***@***.***>; Mention ***@***.***>
Subject: Re: [facebookincubator/velox] Fix seek of PagedInputStream over uncompressed data (PR #1798)

 

@mbasmanova approved this pull request.

@oerling <https://github.com/oerling>  Orri, thank you for fixing this issue. Some comments below.

  _____  

In velox/dwio/dwrf/common/ByteRLE.cpp <https://github.com/facebookincubator/velox/pull/1798#discussion_r894150753> :

> @@ -354,6 +354,8 @@ void ByteRleDecoder::seekToRowGroup(PositionProvider& positionProvider) {
   bufferEnd = bufferStart;
   // force reading a new header
   remainingValues = 0;
+  static int counter = 0;
+  LOG(INFO) << "Suspect Seek #" << counter++;

Should this be removed?

  _____  

In velox/dwio/dwrf/common/PagedInputStream.cpp <https://github.com/facebookincubator/velox/pull/1798#discussion_r894151089> :

> @@ -188,6 +190,16 @@ void PagedInputStream::BackUp(int32_t count) {
       outputBufferPtr_ != nullptr,
       "Backup without previous Next in ",
       getName());
+  if (state_ == State::ORIGINAL) {
+    VELOX_CHECK(

nit: perhaps, split this into two checks so that we have a better idea which boundary is violated in case of failure

  _____  

In velox/dwio/dwrf/common/PagedInputStream.h <https://github.com/facebookincubator/velox/pull/1798#discussion_r894151442> :

>      DWIO_ENSURE(
         decompressor_ || decrypter_,
         "one of decompressor or decryptor is required");
+    LOG(INFO) << "Make " << counter++ << " = " << this;

Remove.

  _____  

In velox/dwio/dwrf/common/PagedInputStream.h <https://github.com/facebookincubator/velox/pull/1798#discussion_r894151630> :

> @@ -125,7 +127,12 @@ class PagedInputStream : public SeekableInputStream {
   size_t remainingLength_{0};
 
   // the last buffer returned from the input

Remove this comment?

  _____  

In velox/dwio/dwrf/common/PagedInputStream.h <https://github.com/facebookincubator/velox/pull/1798#discussion_r894151683> :

> @@ -125,7 +127,12 @@ class PagedInputStream : public SeekableInputStream {
   size_t remainingLength_{0};
 
   // the last buffer returned from the input
+  // The first byte. in the range from input_->Next()

typos: "byte. in the"

—
Reply to this email directly, view it on GitHub <https://github.com/facebookincubator/velox/pull/1798#pullrequestreview-1002244588> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKPPPT77PNWPL5CUDUYITQTVOLDLPANCNFSM5YMK2ZZQ> .
You are receiving this because you were mentioned.Message ID: ***@***.***>
